### PR TITLE
feat: debug block

### PR DESCRIPTION
**Pull request by greged93**
Time spent on this PR: 3 day

Resolves: #1010

# Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Testing

# What is the new behavior?
Adds the `debugTraceBlockByNumber` and `debugTraceBlockByHash` endpoints.

# Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: src/tracing/builder.rs [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+        let mut cfg = CfgEnv::default();
+        cfg.chain_id = eth_provider.chain_id().await?.unwrap_or_default().to();
+
+        let env = Env { cfg, ..Default::default() };
+
+        Ok(Self { eth_provider, env, block: None, _phantom: std::marker::PhantomData })
+    }
+
+    /// Sets the block to trace
+    pub async fn with_block_id(self, block_id: BlockId) -> TracerResult<TracerBuilder<P, Pinned>> {
```
[the remark this answers is not in the document]
Comment (greged93) [reply src/tracing/builder.rs:36]: I don't think this works actually, syntax would be `Self` not `Self<P, Pinned>` which doesn't seem to compile. I think `Self` is an alias for `TracerBuilder<P, Floating>` which isn't what I am returning here

File: src/tracing/builder.rs [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+        let maybe_block = self.block(block_id).await?;
+
+        Ok(TracerBuilder {
+            eth_provider: self.eth_provider.clone(),
+            env: self.env.clone(),
+            block: maybe_block,
+            _phantom: std::marker::PhantomData,
+        })
```
[the remark this answers is not in the document]
Comment (greged93) [reply src/tracing/builder.rs:44]: same issue

File: src/tracing/builder.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (tcoratger) [reply src/tracing/builder.rs:36]: Ah yes true, misread this...

**Review by tcoratger — APPROVED** [review]

File: src/eth_rpc/servers/debug_rpc.rs [hunk 3]
```diff
@@ -141,4 +146,38 @@ impl<P: EthereumProvider + Send + Sync + 'static> DebugApiServer for DebugRpc<P>
...
+    /// Returns the Geth debug trace for the given block number.
+    async fn trace_block_by_number(
+        &self,
+        block_number: BlockNumberOrTag,
+        opts: Option<GethDebugTracingOptions>,
+    ) -> Result<Option<Vec<TraceResult>>> {
+        let provider = Arc::new(&self.eth_provider);
+        let maybe_tracer =
+            TracerBuilder::new(provider).await?.with_block_id(BlockId::Number(block_number)).await?.build()?;
+        if maybe_tracer.is_none() {
```
Comment (Eikix) [inline src/eth_rpc/servers/debug_rpc.rs:159]: why are we doing this here?

why not unwrap_or(Ok(none)) or match?

File: src/eth_rpc/servers/debug_rpc.rs [hunk 4]
```diff
@@ -141,4 +146,38 @@ impl<P: EthereumProvider + Send + Sync + 'static> DebugApiServer for DebugRpc<P>
...
+    /// Returns the Geth debug trace for the given block hash.
+    async fn trace_block_by_hash(
+        &self,
+        block_hash: B256,
+        opts: Option<GethDebugTracingOptions>,
+    ) -> Result<Option<Vec<TraceResult>>> {
+        let provider = Arc::new(&self.eth_provider);
+        let maybe_tracer =
+            TracerBuilder::new(provider).await?.with_block_id(BlockId::Hash(block_hash.into())).await?.build()?;
+        if maybe_tracer.is_none() {
```
Comment (Eikix) [inline src/eth_rpc/servers/debug_rpc.rs:176]: same question

File: src/eth_rpc/servers/debug_rpc.rs (hunk 3 above)
Comment (greged93) [reply src/eth_rpc/servers/debug_rpc.rs:159]: Not sure how you would use unwrap_or and yes we could use a match, just find this more readable

Comment (Eikix) [reply src/eth_rpc/servers/debug_rpc.rs:159]:
```
maybe_tracer
    .map(|tracer| {
        tracer.debug_block(opts.unwrap_or_default())
    })
    .transpose()
```

Comment (greged93) [reply src/eth_rpc/servers/debug_rpc.rs:159]: mmmm I find my version more readable no? can change if you want but this won't change anything performance wise

File: src/tracing/builder.rs [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    eth_provider: P,
+    env: Env,
+    block: Option<reth_rpc_types::Block>,
+    _phantom: std::marker::PhantomData<Status>,
+}
+
+impl<P: EthereumProvider + Send + Sync + Clone> TracerBuilder<P, Floating> {
+    pub async fn new(eth_provider: P) -> TracerResult<Self> {
+        let mut cfg = CfgEnv::default();
+        cfg.chain_id = eth_provider.chain_id().await?.unwrap_or_default().to();
```
[the remark this answers is not in the document]
Comment (greged93) [reply src/tracing/builder.rs:28]: can be sync indeed!

[the remark this answers is not in the document]
Comment (greged93) [reply src/tracing/builder.rs:28]: mmmm it's actually from a trait, what do you think of keeping as is? but yes in fact in won't ever fail and will be sync for the current implementation

[the remark this answers is not in the document]
Comment (Eikix) [reply src/tracing/builder.rs:28]: the trait is imported?

[the remark this answers is not in the document]
Comment (greged93) [reply src/tracing/builder.rs:28]: Not imported but based on Reth

[the remark this answers is not in the document]
Comment (Eikix) [reply src/tracing/builder.rs:28]: If we don't import the trait we could modify the trait no?

[the remark this answers is not in the document]
Comment (greged93) [reply src/tracing/builder.rs:28]: If we never expect any implementation of the trait to be async then we could probably update the trait. If we thing that we might have an implementation async then would leave (also doesn't change much tbh)

**Pull request merged**